File: lib/haveapi/validators/presence.rb
module HaveAPI
  # Checks the value is a number or a string containing only digits.
  #
  # Short form:
  #   string :param, required: true
  #
  # Full form:
  #   string :param, required: {
  #     empty: true/false,
  #     message: 'the error message'
  #   }
  class Validators::Presence < Validator
    name :present
    takes :required, :present

    def setup
      @empty = take(:empty, false)
      @message = take(
          :message,
          @empty ? 'must be present' : 'must be present and non-empty'
      )
    end

    def describe
      {
          empty: @empty,
          message: @message,
      }
    end

    def valid?(v)
      return false if v.nil?
      return !v.strip.empty? if !@empty && v.is_a?(::String)
      # FIXME: other data types?
      true
    end
  end
end

File: lib/haveapi/validator/presence.ex
defmodule HaveAPI.Validator.Presence do
  use HaveAPI.Validator

  def name, do: :present

  def init(opts) do
    empty = Keyword.get(opts, :empty, false)

    %{
      empty: empty,
      message: Keyword.get(
        opts,
        :message,
        (if empty, do: "must be present", else: "must be present and non-empty")
      )
    }
  end

  def validate(%{empty: true}, _v, _params), do: :ok
  def validate(opts, v, _params) when is_binary(v), do: not_empty(opts, String.trim(v))
  def validate(opts, v, _params), do: not_empty(opts, v)

  defp not_empty(opts, nil), do: {:error, [opts.message]}
  defp not_empty(opts, ""), do: {:error, [opts.message]}
  defp not_empty(_opts, _v), do: :ok
end
